### PR TITLE
Clean-up of GPU directives.

### PR DESCRIPTION
**Pull request by johnmauff**
This PR is a clean-up of some of the GPU directives that were causing issues.  I have also enabled a higher level of compiler optimization in the CMake configuration.  This appears to give the same answer on the GPU using the Beltrami test case.  This PR also eliminates the need to build and run Samurai from the bin directory.

### CODE REVIEW

**Review by sjsprecious — COMMENTED** [review]
Thanks @johnmauff for working on this PR. I just have a few minor comments.

File: src/CostFunction.cpp [hunk 1]
```diff
@@ -112,11 +112,11 @@ void CostFunction::truncatedNewton(real* qstate, real* g, const real ftol)
   real *p = new real[nState];
   real *Ap = new real[nState];
   real *r = new real[nState];
+#pragma acc data create(x[0:nState],p[0:nState],Ap[0:nState],r[0:nState])
```
[the remark this answers is not in the document]
Comment (johnmauff) [reply src/CostFunction.cpp:115]: I would prefer to keep the data movement directives in the code for now.  I had always hoped to eliminate the use of management memory in Samurai

File: src/CostFunction3D.cpp [hunk 2]
```diff
@@ -530,7 +530,9 @@ void CostFunction3D::initState(const int iteration)
   calcInnovation();
 
   // Output the original background field
+  //cout << "CostFunction3D::initState: before outputAnalysis" << "\n";
   outputAnalysis("background", bgState);
+  //cout << "CostFunction3D::initState: after outputAnalysis" << "\n";
```
[the remark this answers is not in the document]
Comment (johnmauff) [reply src/CostFunction3D.cpp]: Done

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -209,27 +209,27 @@ if(CMAKE_CXX_COMPILER_ID STREQUAL "NVHPC")
...
     if(MODE STREQUAL "GPU")
-      set_source_files_properties(CostFunction3D.cpp CostFunction.cpp RecursiveFilter.cpp PROPERTIES COMPILE_FLAGS "-O0 -fast ${OpenACC_CXX_FLAGS} -gpu=${OpenACC_ACCEL_TARGET}")
+      set_source_files_properties(CostFunction3D.cpp CostFunction.cpp RecursiveFilter.cpp PROPERTIES COMPILE_FLAGS "-O3 -fast ${OpenACC_CXX_FLAGS} -gpu=${OpenACC_ACCEL_TARGET}")
       target_link_libraries(${PROJECT_NAME} -gpu=${OpenACC_ACCEL_TARGET})
     else()
-      set_source_files_properties(CostFunction3D.cpp RecursiveFilter.cpp PROPERTIES COMPILE_FLAGS "-O0 -fast ${OpenMP_CXX_FLAGS}")
+      set_source_files_properties(CostFunction3D.cpp RecursiveFilter.cpp PROPERTIES COMPILE_FLAGS "-O3 -fast ${OpenMP_CXX_FLAGS}")
     endif()
-    set_source_files_properties(BSplineD.cpp PROPERTIES COMPILE_FLAGS "-O0") # BSpineD.cpp has issues above -O1 with PGI 19.3 & 19.9 at least
+    set_source_files_properties(BSplineD.cpp PROPERTIES COMPILE_FLAGS "-O1") # BSpineD.cpp has issues above -O1 with PGI 19.3 & 19.9 at least
```
[the remark this answers is not in the document]
Comment (johnmauff) [reply src/CMakeLists.txt]: Done

File: src/CostFunction.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjsprecious) [reply src/CostFunction.cpp:115]: That makes sense. Thanks for the clarification.

File: bin/ncar_derecho_build.sh [hunk 4]
```diff
@@ -7,11 +7,11 @@
 
 MODE=${1^^}
 if [ "$MODE" == "GPU" ]; then
-  sed -i 's/MODE CPU/MODE GPU/g' ../CMakeLists.txt
+  sed -i 's/MODE CPU/MODE GPU/g' CMakeLists.txt
```
Comment (sjsprecious) [inline bin/ncar_derecho_build.sh:10]: If you plan to run this script under the root directory, please update the comment at line 3 for clarification. Also do you want to update `ncar_derecho_run.sh` script as well?

Comment (johnmauff) [reply bin/ncar_derecho_build.sh:10]: Good point, I missed the fact that it was being run in the build directory.  Being run from the main directory prevents the need to go searching for the output files, which I spent an embarrassingly large amount of time on last week.

**Review by sjsprecious — APPROVED** [review]
Thanks @johnmauff for addressing the comments. This PR looks good to me and I could use the latest build and run script to test a GPU case on Derecho. The only minor "defect" here is that all the output files are now saved under the main directory rather than a subfolder.

**Comment by johnmauff** [comment]
I don't think it is ideal to drop all the output files in the main directory either. We need a better solution that places the output files in a directory that makes more sense than either the build or the main directory.  Maybe it already exists in Samurai and we just need to figure out how to control it.

**Pull request merged**